In [5]:
import pandas as pd
pd.options.display.max_columns = 99

In [6]:
loans_chunk_5 = pd.read_csv("loans_2007.csv",chunksize=5)

for chunk in loans_chunk_5:
    loans_first_5 = chunk
    break

loans_first_5
loans_first_5.dtypes
obj_cols = loans_first_5.select_dtypes(include=['object']).columns
loans_first_5

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501  1296599.0     5000.0       5000.0           4975.0   36 months   
1  1077430  1314167.0     2500.0       2500.0           2500.0   60 months   
2  1077175  1313524.0     2400.0       2400.0           2400.0   36 months   
3  1076863  1277178.0    10000.0      10000.0          10000.0   36 months   
4  1075358  1311748.0     3000.0       3000.0           3000.0   60 months   

  int_rate  installment grade sub_grade                 emp_title emp_length  \
0   10.65%       162.87     B        B2                       NaN  10+ years   
1   15.27%        59.83     C        C4                     Ryder   < 1 year   
2   15.96%        84.33     C        C5                       NaN  10+ years   
3   13.49%       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4   12.69%        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     24000.0            Verified  Dec-2011   Fully Paid   
1           RENT     30000.0     Source Verified  Dec-2011  Charged Off   
2           RENT     12252.0        Not Verified  Dec-2011   Fully Paid   
3           RENT     49200.0     Source Verified  Dec-2011   Fully Paid   
4           RENT     80000.0     Source Verified  Dec-2011      Current   

  pymnt_plan         purpose                 title zip_code addr_state    dti  \
0          n     credit_card              Computer    860xx         AZ  27.65   
1          n             car                  bike    309xx         GA   1.00   
2          n  small_business  real estate business    606xx         IL   8.72   
3          n           other              personel    917xx         CA  20.00   
4          n           other              Personal    972xx         OR  17.94   

   delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  pub_rec  revol_bal  \
0          0.0         Jan-1985             1.0       3.0      0.0    13648.0   
1          0.0         Apr-1999             5.0       3.0      0.0     1687.0   
2          0.0         Nov-2001             2.0       2.0      0.0     2956.0   
3          0.0         Feb-1996             1.0      10.0      0.0     5598.0   
4          0.0         Jan-1996             0.0      15.0      0.0    27783.0   

  revol_util  total_acc initial_list_status  out_prncp  out_prncp_inv  \
0      83.7%        9.0                   f       0.00           0.00   
1       9.4%        4.0                   f       0.00           0.00   
2      98.5%       10.0                   f       0.00           0.00   
3        21%       37.0                   f       0.00           0.00   
4      53.9%       38.0                   f     461.73         461.73   

    total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
0   5863.155187          5833.84          5000.00         863.16   
1   1008.710000          1008.71           456.46         435.17   
2   3005.666844          3005.67          2400.00         605.67   
3  12231.890000         12231.89         10000.00        2214.92   
4   3581.120000          3581.12          2538.27        1042.85   

   total_rec_late_fee  recoveries  collection_recovery_fee last_pymnt_d  \
0                0.00        0.00                     0.00     Jan-2015   
1                0.00      117.08                     1.11     Apr-2013   
2                0.00        0.00                     0.00     Jun-2014   
3               16.97        0.00                     0.00     Jan-2015   
4                0.00        0.00                     0.00     Jun-2016   

   last_pymnt_amnt last_credit_pull_d  collections_12_mths_ex_med  \
0           171.62           Jun-2016                         0.0   
1           119.66           Sep-2013                         0.0   
2           649.91           Jun-2016                         0.0   
3           357.48           Apr-2016                     

In [7]:
# Memory usage. 
loans_chunk_1000 = pd.read_csv("loans_2007.csv",chunksize=1000)
amt_mem_1000_rows = 0
for chunk in loans_chunk_1000:
    loans_first_1000 = chunk
    amt_mem_1000_rows += chunk.memory_usage(deep=True).sum()
    
print(amt_mem_1000_rows/(1024*1024), "Megabytes")

66.11272621154785 Megabytes


In [8]:
# Types of every column in the data frame. 
loans_first_1000.dtypes.value_counts()

float64    30
object     22
dtype: int64

In [9]:
loans_chunk_1000 = pd.read_csv("loans_2007.csv",chunksize=1000)
total_rows=0
for chunk in loans_chunk_1000:
    total_rows += chunk.shape[0]
print("Total rows: ",total_rows)
print("Total columns: ",30+21+1)



Total rows:  42538
Total columns:  52


In [10]:
# How many unique values are in every object column?

obj_cols_with_less_50perc_unique = []
half_rows = 42538 / 2

for col in obj_cols: 
    list_of_parts_of_column = []
    # reset the iterator. 
    loans_chunk_1000 = pd.read_csv("loans_2007.csv",chunksize=1000)
    
    for chunk in loans_chunk_1000: 
        serie = chunk[col]
        list_of_parts_of_column.append(serie)
    complete_column = pd.concat(list_of_parts_of_column) 
    unique_values = len(complete_column.unique()) 
    if unique_values < half_rows:
               
        obj_cols_with_less_50perc_unique.append((col,unique_values))


In [11]:
# We have here the every object column with 
# his amount of unique values
obj_cols_with_less_50perc_unique

[('term', 3),
 ('int_rate', 395),
 ('grade', 8),
 ('sub_grade', 36),
 ('emp_length', 12),
 ('home_ownership', 6),
 ('verification_status', 4),
 ('issue_d', 56),
 ('loan_status', 10),
 ('pymnt_plan', 3),
 ('purpose', 15),
 ('title', 21265),
 ('zip_code', 838),
 ('addr_state', 51),
 ('earliest_cr_line', 531),
 ('revol_util', 1120),
 ('initial_list_status', 2),
 ('last_pymnt_d', 104),
 ('last_credit_pull_d', 109),
 ('application_type', 2)]

In [12]:
# Checking of float columns

float_cols = loans_first_5.select_dtypes(include=['float']).columns


In [13]:
# How many unique values are in every object column?

float_cols_with_no_null = []

for col in float_cols: 
    list_of_parts_of_column = []
    # reset the iterator. 
    loans_chunk_1000 = pd.read_csv("loans_2007.csv",chunksize=1000)
    
    for chunk in loans_chunk_1000: 
        serie = chunk[col]
        list_of_parts_of_column.append(serie)
    complete_column = pd.concat(list_of_parts_of_column) 

    if complete_column.isnull().sum() == 0:
               
        float_cols_with_no_null.append(col)

float_cols_with_no_null

[]

##### Since the list results empty, every float columns has null values, then they can't be converted in the integer type.  


##### The next step, optimizing the dataframe by converts to better types

In [14]:
# Which object type column can be numeric ones.
loans_first_5[obj_cols]

term int_rate grade sub_grade                 emp_title emp_length  \
0   36 months   10.65%     B        B2                       NaN  10+ years   
1   60 months   15.27%     C        C4                     Ryder   < 1 year   
2   36 months   15.96%     C        C5                       NaN  10+ years   
3   36 months   13.49%     C        C1       AIR RESOURCES BOARD  10+ years   
4   60 months   12.69%     B        B5  University Medical Group     1 year   

  home_ownership verification_status   issue_d  loan_status pymnt_plan  \
0           RENT            Verified  Dec-2011   Fully Paid          n   
1           RENT     Source Verified  Dec-2011  Charged Off          n   
2           RENT        Not Verified  Dec-2011   Fully Paid          n   
3           RENT     Source Verified  Dec-2011   Fully Paid          n   
4           RENT     Source Verified  Dec-2011      Current          n   

          purpose                 title zip_code addr_state earliest_cr_line  \
0     credit_card              Computer    860xx         AZ         Jan-1985   
1             car                  bike    309xx         GA         Apr-1999   
2  small_business  real estate business    606xx         IL         Nov-2001   
3           other              personel    917xx         CA         Feb-1996   
4           other              Personal    972xx         OR         Jan-1996   

  revol_util initial_list_status last_pymnt_d last_credit_pull_d  \
0      83.7%                   f     Jan-2015           Jun-2016   
1       9.4%                   f     Apr-2013           Sep-2013   
2      98.5%                   f     Jun-2014           Jun-2016   
3        21%                   f     Jan-2015           Apr-2016   
4      53.9%                   f     Jun-2016           Jun-2016   

  application_type  
0       INDIVIDUAL  
1       INDIVIDUAL  
2       INDIVIDUAL  
3       INDIVIDUAL  
4       INDIVIDUAL

###### Candidates of object columns to be numeric: 
- term (taking off the string 'months')
- int_rate (taking off the string '%')
- revol_util (taking off the string '%')

###### Candidates  of object columns to be categoric
- All, because all have less than 50% of unique values. 

In [15]:
go_to_numeric = ['term','int_rate','revol_util']
go_to_categoric = []
for col in obj_cols:
    if col not in go_to_numeric:
        go_to_categoric.append(col)
        
go_to_categoric

['grade',
 'sub_grade',
 'emp_title',
 'emp_length',
 'home_ownership',
 'verification_status',
 'issue_d',
 'loan_status',
 'pymnt_plan',
 'purpose',
 'title',
 'zip_code',
 'addr_state',
 'earliest_cr_line',
 'initial_list_status',
 'last_pymnt_d',
 'last_credit_pull_d',
 'application_type']

In [16]:
# float_cols
total_memory_needs = 0

# column already optimized: id

# We are going to optimize object columns that can be 
# converted in numeric and category ones. Also we can optimize 
# the float columns (automated optimization by pandas.to_numeric)
# to pic the right amount of bits in float to every float column


loans_chunks = pd.read_csv("loans_2007.csv",chunksize=1000)
for chunk in loans_chunks: 
    
    for col in go_to_numeric: 
        if col == 'term':
            chunk[col] = chunk[col].str.replace('months','')
            chunk[col] = chunk[col].str.replace(' ','')
            chunk[col] = pd.to_numeric(chunk[col])
        elif col == 'int_rate':
            chunk[col] = chunk[col].str.replace('%','')
            chunk[col] = pd.to_numeric(chunk[col])
        elif col == 'revol_util': 
            chunk[col] = chunk[col].str.replace('%','')
            chunk[col] = pd.to_numeric(chunk[col])
    
    for col in float_cols: 
        chunk[col] = pd.to_numeric(chunk[col])
    
    for col in go_to_categoric: 
        chunk[col] = chunk[col].astype('category')
    
    total_memory_needs += chunk.memory_usage(deep=True).sum()

print(total_memory_needs/(1024*1024), "Megabytes")
        


23.290706634521484 Megabytes


##### The new total memory needs are 23.3 Mb for the full data frame.